web/src/main/java/com/microsoft/applicationinsights/web/internal/WebRequestTrackingFilter.java

```diff
@@ -75,28 +74,35 @@
      * @throws ServletException Exception that can be thrown from invoking the filters chain.
      */
     public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
-        ApplicationInsightsHttpResponseWrapper response = new ApplicationInsightsHttpResponseWrapper((HttpServletResponse)res);
+
+        //From Servlet API 3.0 onward there is getStatus() method in HttpServletResponse class.
+        HttpServletResponse response = (HttpServletResponse)res;
         setKeyOnTLS(key);
 
         boolean isRequestProcessedSuccessfully = invokeSafeOnBeginRequest(req, response);
 
         try {
             chain.doFilter(req, response);
             invokeSafeOnEndRequest(req, response, isRequestProcessedSuccessfully);
-        } catch (ServletException se) {
-            onException(se, req, response,isRequestProcessedSuccessfully);
-            throw se;
-        } catch (IOException ioe) {
-            onException(ioe, req, response, isRequestProcessedSuccessfully);
-            throw ioe;
-        } catch (RuntimeException re) {
-            onException(re, req, response, isRequestProcessedSuccessfully);
-            throw re;
-        } finally {
+        } catch (ServletException | IOException | RuntimeException ex) {
+
+            //Explicitly set the response to 500 as the above three exceptions converts to 500
+            //later in the the chain inside the catch block of org.apache.catalina.core.StandardWrapperValue class
+            //The call to this method happens after the Filter Chain has been completed and thus response object
+            //can never have this value at this time.
+            response.setStatus(500);
```

Comment:
We can't do this. We must discover what the app is returning to the client.

Reply:
@littleaj if you have better alternatives I am happy to take that. But in all my test I did not found any way to control this behavior. Filter chain gets released way before 500 is set.

Reply:
In case of non-tomcat app servers where some other class will govern setting this value, it would also follow the same patter of setting 500. Now ordering might change but the value of response code should be the same in any app server regardless for this three exceptions.